public/app/plugins/datasource/grafana-pyroscope-datasource/QueryEditor/QueryEditor.tsx

```diff
@@ -115,33 +117,62 @@ function useLabels(
   // and what ever has range as dependency is rerun. So this effectively debounces the queries.
   const unpreciseRange = {
-    to: Math.ceil((range?.to.valueOf() || 0) / 5000) * 5000,
-    from: Math.floor((range?.from.valueOf() || 0) / 5000) * 5000,
+    to: Math.ceil((range?.to.valueOf() || 0) / 10000) * 10000,
+    from: Math.floor((range?.from.valueOf() || 0) / 10000) * 10000,
   };
 
-  const labelsResult = useAsync(() => {
-    return datasource.getLabelNames(query.profileTypeId + query.labelSelector, unpreciseRange.from, unpreciseRange.to);
-  }, [datasource, query.profileTypeId, query.labelSelector, unpreciseRange.to, unpreciseRange.from]);
+  // Transforms user input into a valid label selector including the profile type.
+  // It can optionally remove a label, used to support editing existing label values.
+  const createSelector = (rawInput: string, profileTypeId: string, labelToRemove: string): string => {
+    let labels: string[] = [`__profile_type__=\"${profileTypeId}\"`];
+    let match;
+    while ((match = labelSelectorRegex.exec(rawInput)) !== null) {
+      if (match[1] && match[2]) {
+        if (match[1] === labelToRemove) {
+          continue;
+        }
+        labels.push(`${match[1]}=${match[2]}`);
+      }
+    }
+    return `{${labels.join(',')}}`;
+  };
 
-  // Create a function with range and query already baked in so we don't have to send those everywhere
-  const getLabelValues = useCallback(
-    (label: string) => {
-      return datasource.getLabelValues(
-        query.profileTypeId + query.labelSelector,
-        label,
+  const [labels, setLabels] = useState(() => ['']);
+
+  useEffect(() => {
+    const fetchData = async () => {
+      const labels = await datasource.getLabelNames(
+        createSelector(query.labelSelector, query.profileTypeId, ''),
         unpreciseRange.from,
         unpreciseRange.to
       );
+
+      setLabels(labels);
+    };
+    fetchData();
+  }, [query, unpreciseRange.from, unpreciseRange.to, datasource, setLabels]);
+
+  // Create a function with range and query already baked in, so we don't have to send those everywhere
+  const getLabelValues = useCallback(
+    (label: string) => {
+      let labelSelector = createSelector(query.labelSelector, query.profileTypeId, label);
+      return datasource.getLabelValues(labelSelector, label, unpreciseRange.from, unpreciseRange.to);
     },
-    [query, datasource, unpreciseRange.to, unpreciseRange.from]
+    [datasource, query.labelSelector, query.profileTypeId, unpreciseRange.to, unpreciseRange.from]
   );
 
+  const onChangeDebounced = debounce((value: string) => {
+    if (onChange) {
+      onChange({ ...query, labelSelector: value });
+    }
+  }, 200);
+
   const onLabelSelectorChange = useCallback(
     (value: string) => {
-      onChange({ ...query, labelSelector: value });
+      onChangeDebounced(value);
     },
-    [onChange, query]
+    [onChangeDebounced]
```

Comment:
@grafakus @aocenas 
Debouncing is an improvement over directly calling `onChange(query)`, though there are full page re-renders at the debounce interval (200ms) which didn't happen before (we only changed the query object on "blur" prior to the work here). It doesn't feel slow right now, but it might get worse with future work. 

Do you think I should keep this approach or go back to the previous state (having `onLabelSelectorChange` just deal with the label state and a separate `onBlur` for updating the query object)?

Reply:
Makes sense, thank you!